redis/docs/token_authentication.md

```diff
@@ -0,0 +1,217 @@
+- **Multiple Authentication Flows**: Service principals, managed identities, and custom `TokenCredential` implementations
+- **Configurable Refresh Policies**: Customizable refresh thresholds and retry behavior
+
+## Architecture
+
+### Async-First Design
+- Full async/await support for non-blocking operations
+- Seamless integration with multiplexed connections
+
+### Streaming Credentials Pattern
+1. The `EntraIdCredentialsProvider` implements the `StreamingCredentialsProvider` trait, which allows clients to subscribe to a stream of credentials.
+2. An `EntraIdCredentialsProvider` can be created with one of the public constructors. Each of them creates a specific `TokenCredential` implementation from the `azure_identity` crate.
+3. The `EntraIdCredentialsProvider` starts a background task for token refresh.
+4. The `Client` holds a `StreamingCredentialsProvider` and uses it to authenticate connections.
+5. When a `Client` is created, the provider is attached to it with the `with_credentials_provider` function.
+6. The `EntraIdCredentialsProvider` keeps the current token and provides it to a `Client` when it establishes a new `multiplexed_connection`. Before the connection gets established, it creates a background task, which subscribes for credential updates.
+7. When the token is refreshed, the `EntraIdCredentialsProvider` notifies all subscribers with the new credentials.
+8. When a subscriber receives the new credentials, it uses them to re-authenticate itself.
+
+## Quick Start
+
+### 1. Enable the Feature
+
+Add the `entra-id` feature to your `Cargo.toml`:
+
+```toml
+[dependencies]
+redis = { version = "0.32.7", features = ["entra-id", "tokio-comp"] }
+```
+
+### 2. Basic Usage with DefaultAzureCredential
+
+```rust
+use redis::{Client, EntraIdCredentialsProvider, TokenRefreshConfig};
+
+#[tokio::main]
+async fn main() -> redis::RedisResult<()> {
+    // Create credentials provider using DefaultAzureCredential
+    let mut provider = EntraIdCredentialsProvider::new_default()?;
+    provider.start(TokenRefreshConfig::default());
+    
+    // Create Redis client with credentials provider
+    let client = Client::open("redis://your-redis-instance.com:6380")?
+        .with_credentials_provider(provider);
+    
+    // Use the client to get a multiplexed connection
+    let mut con = client.get_multiplexed_async_connection().await?;
+    redis::cmd("SET")
+        .arg("my_key")
+        .arg(42i32)
+        .exec_async(&mut con)
+        .await?;
+    let result: Option<String> = redis::cmd("GET")
+        .arg("my_key")
+        .query_async(&mut con)
+        .await?;
+
+    Ok(())
+}
+```
+
+## Authentication Flows
+
+### DefaultAzureCredential (Recommended for Development)
+
+The `DefaultAzureCredential` tries multiple credential sources in order:
+
+```rust
+let provider = EntraIdCredentialsProvider::new_default()?;
+```
```

Comment:
all of the code documentation should also (or only) appear in code, where it will be tested by rust doc tests